platforms/core-configuration/model-core/src/main/java/org/gradle/api/internal/provider/AbstractProperty.java

```diff
@@ -263,6 +272,33 @@ protected void discardValue() {
         value = state.implicitValue();
     }
 
+    /**
+     * Discards the convention of this property.
+     */
+    protected void discardConvention() {
+        assertCanMutate();
+        value = state.applyConvention(value, getDefaultConvention());
+    }
+
+    public SupportsConvention setToConvention() {
+        this.value = state.setToConvention();
+        return this;
+    }
+
+    public SupportsConvention setToConventionIfUnset() {
+        if (!isDefaultConvention()) {
```

Comment:
Because collection properties differentiate between empty and no value. Conventions on collection properties start as "no value". For `setToConventionIfUnset()`, we don't want to blindly apply a no-value convention as the new explicit value as that would be similarly to unsetting the explicit value (via `value(nul)` or `unset()`). That is exactly what `setToConvention()` does, btw.

The specs for `SupportsConvention.setToConvention()` and `SupportsConvention.setToConventionIfUnset()` (which I will be removing shortly) describe that difference in behavior.